components/raftstore/src/store/fsm/peer.rs

```diff
@@ -2627,6 +2628,33 @@ where
         self.fsm.hibernate_state.count_vote(from.get_id());
     }
 
+    fn on_trace_peer_availability_info(&mut self, from: &metapb::Peer, msg: &ExtraMessage) {
+        if !self.fsm.peer.is_leader() {
+            let mut resp = ExtraMessage::default();
+            resp.set_type(ExtraMessageType::MsgTracePeerAvailabilityInfo);
+            resp.miss_data = self.fsm.peer.miss_data;
+            self.fsm
+                .peer
+                .send_extra_message(resp, &mut self.ctx.trans, from);
+            info!(
+                "peer responses availability info to leader";
+                "region_id" => self.region().get_id(),
+                "peer_id" => self.fsm.peer.peer.get_id(),
+                "leader_id" => from.id,
+            );
+            return;
+        }
+        if !msg.miss_data {
+            self.fsm.peer.peers_miss_data.remove(&from.get_id());
```

Comment:
I'm confused. What's the difference between "peer_id not in peers_miss_data` and `peers_miss_data[peer_id] is false`?

Reply:
Currently, in `on_ready_change_peer`, only `witness` will be put in `peers_miss_data`: 


```rust
if self.fsm.peer.is_leader() {
        if peer.is_witness {
                self.fsm.peer.peers_miss_data.insert(peer.id, false);
        } else peer_prev_is_witness {
                self.fsm.peer.peers_miss_data.insert(peer.id, true);
                self.register_track_non_witnesses_availability_tick();
        }
}
```

So "peer_id not in peers_miss_data" means it's not a witness.

Reply:
That's even more confusing. I think for `is_witness`, it must miss data. But the code says it doesn't.

Reply:
> That's even more confusing. I think for is_witness, it must miss data. But the code says it doesn't.
Indeed, how about

``` rust
if self.fsm.peer.is_leader() {
        if peer.is_witness {
                self.fsm.peer.peers_miss_data.insert(peer.id, true);
        } else peer_prev_is_witness {
                self.fsm.peer.peers_miss_data.insert(peer.id, false);
                self.register_track_non_witnesses_availability_tick();
        }
}
```

then collect pending peers, if it's in peers_miss_data & it's non-witness & miss_data is true, add it into pending_peers.

So "peer_id not in peers_miss_data" means a peer never miss data after inited.

Reply:
Why it's set to false if it's witness in the past? If peers_miss_data[peer_id] is false, one should interpret it as peer_id doesn't miss any data. Then why bother checking the availability?

Reply:
Yes it shouldn't, sorry when I describe

> then collect pending peers, if it's in peers_miss_data & it's non-witness & miss_data is true, then it into pending_peers.

I thought miss_data was still true, didn't realize it was changed to false in the else branch

Reply:
I think what you want to do is check if all voters have data. I suggest to use a `wait_data_peer: Vec<u64>`. If a peer is non-witness, they should be added to the vector. If they have data, then it's removed from the vector.

Reply:
Fixed, thanks!